<a href="https://colab.research.google.com/github/balaram401/Applied-data-science-capstone-project--ibm--coursera-/blob/master/week_4%265_capstone_balu_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [12]:
lon_data=pd.read_csv("https://data.london.gov.uk/download/london-borough-profiles/c1693b82-68b1-44ee-beb2-3decf17dc1f8/london-borough-profiles.csv",encoding= 'unicode_escape')
lon_data.head(20)

Code               Area_name Inner/_Outer_London  \
0   E09000001          City of London        Inner London   
1   E09000002    Barking and Dagenham        Outer London   
2   E09000003                  Barnet        Outer London   
3   E09000004                  Bexley        Outer London   
4   E09000005                   Brent        Outer London   
5   E09000006                 Bromley        Outer London   
6   E09000007                  Camden        Inner London   
7   E09000008                 Croydon        Outer London   
8   E09000009                  Ealing        Outer London   
9   E09000010                 Enfield        Outer London   
10  E09000011               Greenwich        Outer London   
11  E09000012                 Hackney        Inner London   
12  E09000013  Hammersmith and Fulham        Inner London   
13  E09000014                Haringey        Inner London   
14  E09000015                  Harrow        Outer London   
15  E09000016                Havering        Outer London   
16  E09000017              Hillingdon        Outer London   
17  E09000018                Hounslow        Outer London   
18  E09000019               Islington        Inner London   
19  E09000020  Kensington and Chelsea        Inner London   

    GLA_Population_Estimate_2017 GLA_Household_Estimate_2017  \
0                           8800                        5326   
1                         209000                       78188   
2                         389600                      151423   
3                         244300                       97736   
4                         332100                      121048   
5                         327900                      140602   
6                         242500                      107654   
7                         386500                      159010   
8                         351600                      132663   
9                         333000                      130328   
10                        280100                      113964   
11                        274300                      115417   
12                        185300                       83552   
13                        278000                      115608   
14                        252300                       92557   
15                        254300                      104098   
16                        301000                      110827   
17                        274200                      105887   
18                        231200                      105038   
19                        159000                       80200   

   Inland_Area_(Hectares) Population_density_(per_hectare)_2017  \
0                     290                                  30.3   
1                   3,611                                  57.9   
2                   8,675                                  44.9   
3                   6,058                                  40.3   
4                   4,323                                  76.8   
5                  15,013                                  21.8   
6                   2,179                                 111.3   
7                   8,650                                  44.7   
8                   5,554                                  63.3   
9                   8,083                                  41.2   
10                  4,733                                  59.2   
11                  1,905                                   144   
12                  1,640                                   113   
13                  2,960                                  93.9   
14                  5,046                                    50   
15                 11,235                                  22.6   
16                 11,570                                    26   
17                  5,598                                    49   
18                  1,486                                 155.6   
19                  1,212                 

In [16]:
columns=lon_data.columns
columns

Index(['Code', 'Area_name', 'Inner/_Outer_London',
       'GLA_Population_Estimate_2017', 'GLA_Household_Estimate_2017',
       'Inland_Area_(Hectares)', 'Population_density_(per_hectare)_2017',
       'Average_Age,_2017', 'Proportion_of_population_aged_0-15,_2015',
       'Proportion_of_population_of_working-age,_2015',
       'Proportion_of_population_aged_65_and_over,_2015',
       'Net_internal_migration_(2015)', 'Net_international_migration_(2015)',
       'Net_natural_change_(2015)',
       '%_of_resident_population_born_abroad_(2015)',
       'Largest_migrant_population_by_country_of_birth_(2011)',
       '%_of_largest_migrant_population_(2011)',
       'Second_largest_migrant_population_by_country_of_birth_(2011)',
       '%_of_second_largest_migrant_population_(2011)',
       'Third_largest_migrant_population_by_country_of_birth_(2011)',
       '%_of_third_largest_migrant_population_(2011)',
       '%_of_population_from_BAME_groups_(2016)',
       '%_people_aged_3+_whose_main_

In [17]:
lon_data.info

<bound method DataFrame.info of          Code               Area_name Inner/_Outer_London  \
0   E09000001          City of London        Inner London   
1   E09000002    Barking and Dagenham        Outer London   
2   E09000003                  Barnet        Outer London   
3   E09000004                  Bexley        Outer London   
4   E09000005                   Brent        Outer London   
5   E09000006                 Bromley        Outer London   
6   E09000007                  Camden        Inner London   
7   E09000008                 Croydon        Outer London   
8   E09000009                  Ealing        Outer London   
9   E09000010                 Enfield        Outer London   
10  E09000011               Greenwich        Outer London   
11  E09000012                 Hackney        Inner London   
12  E09000013  Hammersmith and Fulham        Inner London   
13  E09000014                Haringey        Inner London   
14  E09000015                  Harrow        Outer Lo

In [36]:

df = pd.DataFrame(columns)

df.tail(84)

0
0                                                Code
1                                           Area_name
2                                 Inner/_Outer_London
3                        GLA_Population_Estimate_2017
4                         GLA_Household_Estimate_2017
5                              Inland_Area_(Hectares)
6               Population_density_(per_hectare)_2017
7                                   Average_Age,_2017
8            Proportion_of_population_aged_0-15,_2015
9       Proportion_of_population_of_working-age,_2015
10    Proportion_of_population_aged_65_and_over,_2015
11                      Net_internal_migration_(2015)
12                 Net_international_migration_(2015)
13                          Net_natural_change_(2015)
14        %_of_resident_population_born_abroad_(2015)
15  Largest_migrant_population_by_country_of_birth...
16             %_of_largest_migrant_population_(2011)
17  Second_largest_migrant_population_by_country_o...
18      %_of_second_largest_migrant_population_(2011)
19  Third_largest_migrant_population_by_country_of...
20       %_of_third_largest_migrant_population_(2011)
21            %_of_population_from_BAME_groups_(2016)
22  %_people_aged_3+_whose_main_language_is_not_En...
23  Overseas_nationals_entering_the_UK_(NINo),_(20...
24                New_migrant_(NINo)_rates,_(2015/16)
25  Largest_migrant_population_arrived_during_2015/16
26  Second_largest_migrant_population_arrived_duri...
27  Third_largest_migrant_population_arrived_durin...
28                         Employment_rate_(%)_(2015)
29                        Male_employment_rate_(2015)
30                      Female_employment_rate_(2015)
31                           Unemployment_rate_(2015)
32  Youth_Unemployment_(claimant)_rate_18-24_(Dec-15)
33  Proportion_of_16-18_year_olds_who_are_NEET_(%)...
34  Proportion_of_the_working-age_population_who_c...
35             %_working-age_with_a_disability_(2015)
36  Proportion_of_working_age_people_with_no_quali...
37  Proportion_of_working_age_with_degree_or_equiv...
38                           Gross_Annual_Pay,_(2016)
39                     Gross_Annual_Pay_-_Male_(2016)
40                   Gross_Annual_Pay_-_Female_(2016)
41  Modelled_Household_median_income_estimates_201...
42  %_adults_that_volunteered_in_past_12_months_(2...
43                 Number_of_jobs_by_workplace_(2014)
44    %_of_employment_that_is_in_public_sector_(2014)
45                                 Jobs_Density,_2015
46                  Number_of_active_businesses,_2015
47  Two-year_business_survival_rates_(started_in_2...
48        Crime_rates_per_thousand_population_2014/15
49               Fires_per_thousand_population_(2014)
50  Ambulance_incidents_per_hundred_population_(2014)
51                           Median_House_Price,_2015
52     Average_Band_D_Council_Tax_charge_(£),_2015/16
53              New_Homes_(net)_2015/16_(provisional)
54                     Homes_Owned_outright,_(2014)_%
55       Being_bought_with_mortgage_or_loan,_(2014)_%
56  Rented_from_Local_Authority_or_Housing_Associa...
57             Rented_from_Private_landlord,_(2014)_%
58                 %_of_area_that_is_Greenspace,_2005
59                      Total_carbon_emissions_(2014)
60            Household_Waste_Recycling_Rate,_2014/15
61                      Number_of_cars,_(2011_Census)
62        Number_of_cars_per_household,_(2011_Census)
63  %_of_adults_who_cycle_at_least_once_per_month,...
64  Average_Public_Transport_Accessibility_score,_...
65  Achievement_of_5_or_more_A*-_C_grades_at_GCSE_...
66              Rates_of_Children_Looked_After_(2016)
67  %_of_pupils_whose_first_language_is_not_Englis...
68  %_children_living_in_out-of-work_households_(2...
69                    Male_life_expectancy,_(2012-14)
70                  Female_life_expectancy,_(2012-14)
71                     Teenage_conception_rate_(2014)
72        Life_satisfaction_score_2011-14_(out_of_10)
73           Worthwhileness_score_2011-14_(out_of_10)
74

In [0]:
#headers =['Code', 'Area_name', 'Inner/_Outer_London','Average_Public_Transport_Accessibility_score,_2014','Life_satisfaction_score_2011-14_(out_of_10)','Worthwhileness_score_2011-14_(out_of_10)','Happiness_score_2011-14_(out_of_10)']


In [52]:
#this is for the clusterring of the area with best living index for a housing corporation !!!

lon_data_new=lon_data[['Code', 'Area_name','Average_Public_Transport_Accessibility_score,_2014','Crime_rates_per_thousand_population_2014/15','Life_satisfaction_score_2011-14_(out_of_10)','Worthwhileness_score_2011-14_(out_of_10)','Happiness_score_2011-14_(out_of_10)']]
lon_data_new.head(20)

Code               Area_name  \
0   E09000001          City of London   
1   E09000002    Barking and Dagenham   
2   E09000003                  Barnet   
3   E09000004                  Bexley   
4   E09000005                   Brent   
5   E09000006                 Bromley   
6   E09000007                  Camden   
7   E09000008                 Croydon   
8   E09000009                  Ealing   
9   E09000010                 Enfield   
10  E09000011               Greenwich   
11  E09000012                 Hackney   
12  E09000013  Hammersmith and Fulham   
13  E09000014                Haringey   
14  E09000015                  Harrow   
15  E09000016                Havering   
16  E09000017              Hillingdon   
17  E09000018                Hounslow   
18  E09000019               Islington   
19  E09000020  Kensington and Chelsea   

   Average_Public_Transport_Accessibility_score,_2014  \
0                                                 7.9   
1                                                   3   
2                                                   3   
3                                                 2.6   
4                                                 3.7   
5                                                 2.8   
6                                                 5.7   
7                                                 3.2   
8                                                 3.3   
9                                                   3   
10                                                3.4   
11                                                4.9   
12                                                4.7   
13                                                4.3   
14                                                2.9   
15                                                2.5   
16                                                2.4   
17                                                  3   
18                                                5.7   
19                                                5.8   

   Crime_rates_per_thousand_population_2014/15  \
0                                            .   
1                                         83.4   
2                                         62.7   
3                                         51.8   
4                                         78.8   
5                                         64.1   
6                                        123.5   
7                                           77   
8                                         75.5   
9                                         69.4   
10                                        79.4   
11                                        99.6   
12                                       113.2   
13                                        90.2   
14                                        50.4   
15                                        62.9   
16                                        76.6   
17                                        79.2   
18                                       121.2   
19                                       120.9   

    Life_satisfaction_score_2011-14_(out_of_10)  \
0                                           6.6   
1                                           7.1   
2                                           7.5   
3                                           7.4   
4                                           7.3   
5                                           7.5   
6                                           7.1   
7                                           7.1   
8                                           7.3   
9                                           7.3   
10                                          7.2   
11                                          7.0   
12                                          7.4   
13                                          7.2   
14                                          7.3   
15                                          7.4   
16                             

In [0]:
#A crime rate is calculated by dividing the number of reported crimes by the total population; the result is multiplied by 100,000. For example, in 2010 there were 58,100 robberies in California and the population was 38,826,898. This equals a robbery crime rate of 149.6 per 100,000 general population.

A crime rate is calculated by dividing the number of reported crimes by the total population; the result is multiplied by 100,000. For example, in 2010 there were 58,100 robberies in California and the population was 38,826,898. This equals a robbery crime rate of 149.6 per 100,000 general population.

In [61]:
#creating a data frame for the best for startung any business .
lon_data_business=lon_data[['Area_name','Unemployment_rate_(2015)','Gross_Annual_Pay,_(2016)','Number_of_active_businesses,_2015','Two-year_business_survival_rates_(started_in_2013)']]
lon_data_business.head(10)

Area_name Unemployment_rate_(2015) Gross_Annual_Pay,_(2016)  \
0        City of London                        .                        .   
1  Barking and Dagenham                       11                    27886   
2                Barnet                      8.5                    33443   
3                Bexley                      7.6                    34350   
4                 Brent                      7.5                    29812   
5               Bromley                      5.3                    37682   
6                Camden                        4                    39796   
7               Croydon                      4.1                    32696   
8                Ealing                      5.8                    31331   
9               Enfield                      3.8                    31603   

   Number_of_active_businesses,_2015  \
0                              26130   
1                               6560   
2                              26190   
3                               9075   
4                              15745   
5                              15695   
6                              31385   
7                              15540   
8                              18700   
9                              13925   

   Two-year_business_survival_rates_(started_in_2013)  
0                                               64.3   
1                                               73.0   
2                                               73.8   
3                                               73.5   
4                                               74.4   
5                                               78.6   
6                                               73.6   
7                                               75.3   
8                                               75.8   
9                                               74.2

In [54]:
#defining the area names.
area_name=lon_data[['Area_name']]
area_name

Area_name
0           City of London
1     Barking and Dagenham
2                   Barnet
3                   Bexley
4                    Brent
5                  Bromley
6                   Camden
7                  Croydon
8                   Ealing
9                  Enfield
10               Greenwich
11                 Hackney
12  Hammersmith and Fulham
13                Haringey
14                  Harrow
15                Havering
16              Hillingdon
17                Hounslow
18               Islington
19  Kensington and Chelsea
20    Kingston upon Thames
21                 Lambeth
22                Lewisham
23                  Merton
24                  Newham
25               Redbridge
26    Richmond upon Thames
27               Southwark
28                  Sutton
29           Tower Hamlets
30          Waltham Forest
31              Wandsworth
32             Westminster
33            Inner London
34            Outer London
35                  London
36                 England
37          United Kingdom

In [48]:
!pip install geocoder
import geocoder
#finding the coordinates of the area names or neighbourhoods in london.
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, London, England'.format(neighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

In [0]:
coords = [ get_latlng(neighborhood) for neighborhood in area_name["Area_name"].tolist() ]

In [50]:
coords

[[51.52050000000003, -0.09742999999997437],
 [51.54393218800004, 0.1331566440000529],
 [51.62729399999999, -0.25375949999999536],
 [51.622832082632826, -0.08065551264637581],
 [51.60977558622067, -0.19467855915694893],
 [51.431820424666206, -0.016565652066046606],
 [51.53236000000004, -0.1279599999999732],
 [51.593209478792595, -0.0833901391886536],
 [51.51406000000003, -0.30072999999993044],
 [51.54002082084554, -0.07750147872409442],
 [51.48454000000004, 0.002750000000048658],
 [51.54505000000006, -0.05531999999993786],
 [51.4826899378211, -0.2129099233942567],
 [51.589264499999985, -0.10640475000000119],
 [51.513180000000034, -0.10697999999996455],
 [51.54460488121069, -0.14410476132484631],
 [51.484225357256605, -0.0964798815487847],
 [51.467220000000054, -0.3665299999999547],
 [51.532790000000034, -0.1061399999999253],
 [51.510380000000055, -0.3314699999999675],
 [51.41124000000008, -0.2910499999999274],
 [51.49084000000005, -0.11107999999995855],
 [51.459160000000054, -0.01212999

In [0]:
# create temporary dataframe to populate the coordinates into Latitude and Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [76]:
#there are two data frames now,london data new and london data business.
#merging the lat and long with the lon_data_new,lon_data_business,area_name.
lon_data_new['Latitude'] = df_coords['Latitude']
lon_data_new['Longitude'] = df_coords['Longitude']
lon_data_new.head(84)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Code               Area_name  \
0   E09000001          City of London   
1   E09000002    Barking and Dagenham   
2   E09000003                  Barnet   
3   E09000004                  Bexley   
4   E09000005                   Brent   
5   E09000006                 Bromley   
6   E09000007                  Camden   
7   E09000008                 Croydon   
8   E09000009                  Ealing   
9   E09000010                 Enfield   
10  E09000011               Greenwich   
11  E09000012                 Hackney   
12  E09000013  Hammersmith and Fulham   
13  E09000014                Haringey   
14  E09000015                  Harrow   
15  E09000016                Havering   
16  E09000017              Hillingdon   
17  E09000018                Hounslow   
18  E09000019               Islington   
19  E09000020  Kensington and Chelsea   
20  E09000021    Kingston upon Thames   
21  E09000022                 Lambeth   
22  E09000023                Lewisham   
23  E09000024                  Merton   
24  E09000025                  Newham   
25  E09000026               Redbridge   
26  E09000027    Richmond upon Thames   
27  E09000028               Southwark   
28  E09000029                  Sutton   
29  E09000030           Tower Hamlets   
30  E09000031          Waltham Forest   
31  E09000032              Wandsworth   
32  E09000033             Westminster   
33  E13000001            Inner London   
34  E13000002            Outer London   
35  E12000007                  London   
36  E92000001                 England   
37  K02000001          United Kingdom   

   Average_Public_Transport_Accessibility_score,_2014  \
0                                                 7.9   
1                                                   3   
2                                                   3   
3                                                 2.6   
4                                                 3.7   
5                                                 2.8   
6                                                 5.7   
7                                                 3.2   
8                                                 3.3   
9                                                   3   
10                                                3.4   
11                                                4.9   
12                                                4.7   
13                                                4.3   
14                                                2.9   
15                                                2.5   
16                                                2.4   
17                                                  3   
18                                                5.7   
19                                                5.8   
20                                                2.9   
21                                                  5   
22                                                4.1   
23                                                3.4   
24                                                3.9   
25                                                  3   
26                                                3.1   
27                                                4.9   
28                                                2.9   
29                                                  5   
30                                                3.6   
31                                                4.3   
32                                                6.5   
33                                                4.9   
34                                                  3   
35                                                3.8   
36                                                  .   
37                                                  .   

   Crime_rates_per_thousand_population_2014/15  \
0                                            .   
1                                         83.4   
2                                  

In [62]:
#business_data _merging 
lon_data_business['Latitude'] = df_coords['Latitude']
lon_data_business['Longitude'] = df_coords['Longitude']
lon_data_business.head()


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Area_name Unemployment_rate_(2015) Gross_Annual_Pay,_(2016)  \
0        City of London                        .                        .   
1  Barking and Dagenham                       11                    27886   
2                Barnet                      8.5                    33443   
3                Bexley                      7.6                    34350   
4                 Brent                      7.5                    29812   

   Number_of_active_businesses,_2015  \
0                              26130   
1                               6560   
2                              26190   
3                               9075   
4                              15745   

   Two-year_business_survival_rates_(started_in_2013)   Latitude  Longitude  
0                                               64.3   51.520500  -0.097430  
1                                               73.0   51.543932   0.133157  
2                                               73.8   51.627294  -0.253759  
3                                               73.5   51.622832  -0.080656  
4                                               74.4   51.609776  -0.194679

In [64]:
#we can use foursquare api for this .
area_name['Latitude'] = df_coords['Latitude']
area_name['Longitude'] = df_coords['Longitude']
area_name.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Area_name   Latitude  Longitude
0        City of London  51.520500  -0.097430
1  Barking and Dagenham  51.543932   0.133157
2                Barnet  51.627294  -0.253759
3                Bexley  51.622832  -0.080656
4                 Brent  51.609776  -0.194679

In [0]:
area_name.to_csv("area_name.csv", index=False)

In [66]:
#creating amap of london with neighbourhoods 
# get the coordinates of londoan
address = 'London, England'

geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of London,England  {}, {}.'.format(latitude, longitude))

The geograpical coordinate of London,England  51.5073219, -0.1276474.


In [68]:
# create map of london using latitude and longitude values
map_london = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, areaname in zip(area_name['Latitude'], area_name['Longitude'], area_name['Area_name']):
    label = '{}'.format(areaname)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_london)  
    
map_london

In [79]:
def convert_fill(lon_data_new):
    return lon_data_new.stack().apply(pd.to_numeric, errors='ignore').fillna(0).unstack()

convert_fill(lon_data_new)

Code               Area_name  \
0   E09000001          City of London   
1   E09000002    Barking and Dagenham   
2   E09000003                  Barnet   
3   E09000004                  Bexley   
4   E09000005                   Brent   
5   E09000006                 Bromley   
6   E09000007                  Camden   
7   E09000008                 Croydon   
8   E09000009                  Ealing   
9   E09000010                 Enfield   
10  E09000011               Greenwich   
11  E09000012                 Hackney   
12  E09000013  Hammersmith and Fulham   
13  E09000014                Haringey   
14  E09000015                  Harrow   
15  E09000016                Havering   
16  E09000017              Hillingdon   
17  E09000018                Hounslow   
18  E09000019               Islington   
19  E09000020  Kensington and Chelsea   
20  E09000021    Kingston upon Thames   
21  E09000022                 Lambeth   
22  E09000023                Lewisham   
23  E09000024                  Merton   
24  E09000025                  Newham   
25  E09000026               Redbridge   
26  E09000027    Richmond upon Thames   
27  E09000028               Southwark   
28  E09000029                  Sutton   
29  E09000030           Tower Hamlets   
30  E09000031          Waltham Forest   
31  E09000032              Wandsworth   
32  E09000033             Westminster   
33  E13000001            Inner London   
34  E13000002            Outer London   
35  E12000007                  London   
36  E92000001                 England   
37  K02000001          United Kingdom   

   Average_Public_Transport_Accessibility_score,_2014  \
0                                                 7.9   
1                                                   3   
2                                                   3   
3                                                 2.6   
4                                                 3.7   
5                                                 2.8   
6                                                 5.7   
7                                                 3.2   
8                                                 3.3   
9                                                   3   
10                                                3.4   
11                                                4.9   
12                                                4.7   
13                                                4.3   
14                                                2.9   
15                                                2.5   
16                                                2.4   
17                                                  3   
18                                                5.7   
19                                                5.8   
20                                                2.9   
21                                                  5   
22                                                4.1   
23                                                3.4   
24                                                3.9   
25                                                  3   
26                                                3.1   
27                                                4.9   
28                                                2.9   
29                                                  5   
30                                                3.6   
31                                                4.3   
32                                                6.5   
33                                                4.9   
34                                                  3   
35                                                3.8   
36                                                  .   
37                                                  .   

   Crime_rates_per_thousand_population_2014/15  \
0                                            .   
1                                         83.4   
2                                  

In [90]:
#replacing all the dot values into 0 ;;;;
lon_data_new=lon_data_new[['Code', 'Area_name','Average_Public_Transport_Accessibility_score,_2014','Crime_rates_per_thousand_population_2014/15','Life_satisfaction_score_2011-14_(out_of_10)','Worthwhileness_score_2011-14_(out_of_10)','Happiness_score_2011-14_(out_of_10)']].replace('.',0)
lon_data_new.tail()

Code       Area_name  \
33  E13000001    Inner London   
34  E13000002    Outer London   
35  E12000007          London   
36  E92000001         England   
37  K02000001  United Kingdom   

   Average_Public_Transport_Accessibility_score,_2014  \
33                                                4.9   
34                                                  3   
35                                                3.8   
36                                                  0   
37                                                  0   

   Crime_rates_per_thousand_population_2014/15  \
33                                       106.4   
34                                        69.4   
35                                          84   
36                                        65.7   
37                                           0   

    Life_satisfaction_score_2011-14_(out_of_10)  \
33                                          7.3   
34                                          7.3   
35                                          7.3   
36                                          7.5   
37                                          7.5   

    Worthwhileness_score_2011-14_(out_of_10)  \
33                                       7.5   
34                                       7.6   
35                                       7.6   
36                                       7.7   
37                                       7.7   

    Happiness_score_2011-14_(out_of_10)  
33                                  7.2  
34                                  7.3  
35                                  7.2  
36                                  7.3  
37                                  7.3

In [94]:
#clusterin the data with k maps .
num_clusters = 3
lon_drop=lon_data_new.drop(['Code','Area_name'],1)
k_means = KMeans(init="k-means++", n_clusters=num_clusters, n_init=12)
k_means.fit(lon_drop)
labels = k_means.labels_

print(labels)

[0 0 0 0 0 0 1 0 0 0 0 1 1 1 0 0 0 0 1 1 0 1 0 0 1 0 0 1 0 1 0 0 2 1 0 0 0
 0]


In [127]:
#here is the dataframe with cluster data ......................................................................
lon_data_new["Labels"] = labels
lon_data_new.head(5)
lon_data_new['Latitude'] = df_coords['Latitude']
lon_data_new['Longitude'] = df_coords['Longitude']
lon_data_new.head()

Code             Area_name  \
0  E09000001        City of London   
1  E09000002  Barking and Dagenham   
2  E09000003                Barnet   
3  E09000004                Bexley   
4  E09000005                 Brent   

  Average_Public_Transport_Accessibility_score,_2014  \
0                                                7.9   
1                                                  3   
2                                                  3   
3                                                2.6   
4                                                3.7   

  Crime_rates_per_thousand_population_2014/15  \
0                                           0   
1                                        83.4   
2                                        62.7   
3                                        51.8   
4                                        78.8   

   Life_satisfaction_score_2011-14_(out_of_10)  \
0                                          6.6   
1                                          7.1   
2                                          7.5   
3                                          7.4   
4                                          7.3   

   Worthwhileness_score_2011-14_(out_of_10)  \
0                                       7.1   
1                                       7.6   
2                                       7.8   
3                                       7.7   
4                                       7.4   

   Happiness_score_2011-14_(out_of_10)  Labels   Latitude  Longitude  
0                                  6.0       0  51.520500  -0.097430  
1                                  7.1       0  51.543932   0.133157  
2                                  7.4       0  51.627294  -0.253759  
3                                  7.2       0  51.622832  -0.080656  
4                                  7.2       0  51.609776  -0.194679

In [101]:
#To check out the centroid values by averaging the values in each clustter .
lon_data_mean=lon_data_new.groupby('Labels').mean()
lon_data_mean

Life_satisfaction_score_2011-14_(out_of_10)  \
Labels                                                
0                                          7.307692   
1                                          7.245455   
2                                          7.300000   

        Worthwhileness_score_2011-14_(out_of_10)  \
Labels                                             
0                                       7.611538   
1                                       7.490909   
2                                       7.500000   

        Happiness_score_2011-14_(out_of_10)  
Labels                                       
0                                  7.223077  
1                                  7.209091  
2                                  7.100000

In [162]:
lon_data_mean_new=lon_data_mean.mean(axis=1)
lon_data_mean_new.to_frame()


0
Labels          
0       7.380769
1       7.315152
2       7.300000

In [130]:
lon_data_new.sort_values(["Labels"], inplace=True)
lon_data_new.head(37)

Code               Area_name  \
0   E09000001          City of London   
26  E09000027    Richmond upon Thames   
20  E09000021    Kingston upon Thames   
22  E09000023                Lewisham   
23  E09000024                  Merton   
25  E09000026               Redbridge   
35  E12000007                  London   
28  E09000029                  Sutton   
31  E09000032              Wandsworth   
34  E13000002            Outer London   
17  E09000018                Hounslow   
36  E92000001                 England   
10  E09000011               Greenwich   
30  E09000031          Waltham Forest   
15  E09000016                Havering   
16  E09000017              Hillingdon   
14  E09000015                  Harrow   
7   E09000008                 Croydon   
4   E09000005                   Brent   
8   E09000009                  Ealing   
5   E09000006                 Bromley   
3   E09000004                  Bexley   
2   E09000003                  Barnet   
1   E09000002    Barking and Dagenham   
37  K02000001          United Kingdom   
9   E09000010                 Enfield   
18  E09000019               Islington   
13  E09000014                Haringey   
33  E13000001            Inner London   
29  E09000030           Tower Hamlets   
6   E09000007                  Camden   
19  E09000020  Kensington and Chelsea   
11  E09000012                 Hackney   
12  E09000013  Hammersmith and Fulham   
27  E09000028               Southwark   
24  E09000025                  Newham   
21  E09000022                 Lambeth   

   Average_Public_Transport_Accessibility_score,_2014  \
0                                                 7.9   
26                                                3.1   
20                                                2.9   
22                                                4.1   
23                                                3.4   
25                                                  3   
35                                                3.8   
28                                                2.9   
31                                                4.3   
34                                                  3   
17                                                  3   
36                                                  0   
10                                                3.4   
30                                                3.6   
15                                                2.5   
16                                                2.4   
14                                                2.9   
7                                                 3.2   
4                                                 3.7   
8                                                 3.3   
5                                                 2.8   
3                                                 2.6   
2                                                   3   
1                                                   3   
37                                                  0   
9                                                   3   
18                                                5.7   
13                                                4.3   
33                                                4.9   
29                                                  5   
6                                                 5.7   
19                                                5.8   
11                                                4.9   
12                                                4.7   
27                                                4.9   
24                                                3.9   
21                                                  5   

   Crime_rates_per_thousand_population_2014/15  \
0                                            0   
26                                        56.3   
20                                        58.5   
22                                          77   
23                               

In [132]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(num_clusters)
ys = [i+x+(i*x)**2 for i in range(num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(lon_data_new['Latitude'], lon_data_new['Longitude'], lon_data_new['Area_name'], lon_data_new['Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [0]:
#===============================================

In [165]:
lon_data_new.loc[kl_merged['Cluster Labels'] == 0]

Code               Area_name  \
20  E09000021    Kingston upon Thames   
22  E09000023                Lewisham   
23  E09000024                  Merton   
35  E12000007                  London   
30  E09000031          Waltham Forest   
15  E09000016                Havering   
14  E09000015                  Harrow   
8   E09000009                  Ealing   
3   E09000004                  Bexley   
2   E09000003                  Barnet   
9   E09000010                 Enfield   
33  E13000001            Inner London   
6   E09000007                  Camden   
12  E09000013  Hammersmith and Fulham   
27  E09000028               Southwark   
24  E09000025                  Newham   
21  E09000022                 Lambeth   

   Average_Public_Transport_Accessibility_score,_2014  \
20                                                2.9   
22                                                4.1   
23                                                3.4   
35                                                3.8   
30                                                3.6   
15                                                2.5   
14                                                2.9   
8                                                 3.3   
3                                                 2.6   
2                                                   3   
9                                                   3   
33                                                4.9   
6                                                 5.7   
12                                                4.7   
27                                                4.9   
24                                                3.9   
21                                                  5   

   Crime_rates_per_thousand_population_2014/15  \
20                                        58.5   
22                                          77   
23                                        59.7   
35                                          84   
30                                          78   
15                                        62.9   
14                                        50.4   
8                                         75.5   
3                                         51.8   
2                                         62.7   
9                                         69.4   
33                                       106.4   
6                                        123.5   
12                                       113.2   
27                                       100.6   
24                                        90.8   
21                                       104.6   

    Life_satisfaction_score_2011-14_(out_of_10)  \
20                                          7.3   
22                                          7.2   
23                                          7.3   
35                                          7.3   
30                                          7.2   
15                                          7.4   
14                                          7.3   
8                                           7.3   
3                                           7.4   
2                                           7.5   
9                                           7.3   
33                                          7.3   
6                                           7.1   
12                                          7.4   
27                                          7.5   
24                                          7.1   
21                                          7.1   

    Worthwhileness_score_2011-14_(out_of_10)  \
20                                       7.7   
22                                       7.6   
23                                       7.5   
35                                       7.6   
30                                       7.5   
15                                       7.7   
14                                       7.4   
8                                      

In [166]:
lon_data_new.loc[kl_merged['Cluster Labels'] == 1]

Code               Area_name  \
0   E09000001          City of London   
25  E09000026               Redbridge   
28  E09000029                  Sutton   
31  E09000032              Wandsworth   
34  E13000002            Outer London   
10  E09000011               Greenwich   
16  E09000017              Hillingdon   
4   E09000005                   Brent   
5   E09000006                 Bromley   
1   E09000002    Barking and Dagenham   
37  K02000001          United Kingdom   
18  E09000019               Islington   
19  E09000020  Kensington and Chelsea   
32  E09000033             Westminster   

   Average_Public_Transport_Accessibility_score,_2014  \
0                                                 7.9   
25                                                  3   
28                                                2.9   
31                                                4.3   
34                                                  3   
10                                                3.4   
16                                                2.4   
4                                                 3.7   
5                                                 2.8   
1                                                   3   
37                                                  0   
18                                                5.7   
19                                                5.8   
32                                                6.5   

   Crime_rates_per_thousand_population_2014/15  \
0                                            0   
25                                        69.7   
28                                        55.9   
31                                        72.6   
34                                        69.4   
10                                        79.4   
16                                        76.6   
4                                         78.8   
5                                         64.1   
1                                         83.4   
37                                           0   
18                                       121.2   
19                                       120.9   
32                                       212.4   

    Life_satisfaction_score_2011-14_(out_of_10)  \
0                                           6.6   
25                                          7.3   
28                                          7.5   
31                                          7.4   
34                                          7.3   
10                                          7.2   
16                                          7.4   
4                                           7.3   
5                                           7.5   
1                                           7.1   
37                                          7.5   
18                                          7.2   
19                                          7.6   
32                                          7.3   

    Worthwhileness_score_2011-14_(out_of_10)  \
0                                        7.1   
25                                       7.6   
28                                       7.7   
31                                       7.6   
34                                       7.6   
10                                       7.6   
16                                       7.7   
4                                        7.4   
5                                        7.9   
1                                        7.6   
37                                       7.7   
18                                       7.4   
19                                       7.8   
32                                       7.5   

    Happiness_score_2011-14_(out_of_10)  Labels   Latitude  Longitude  
0                                   6.0       0  51.520500  -0.097430  
25                                  7.3       0  51.475778  -0.080698  
28                                  7.3       0  51.490987  -0.167417  
31         

In [167]:

lon_data_new.loc[kl_merged['Cluster Labels'] == 2]


Code             Area_name  \
26  E09000027  Richmond upon Thames   
17  E09000018              Hounslow   
36  E92000001               England   
7   E09000008               Croydon   
13  E09000014              Haringey   
29  E09000030         Tower Hamlets   
11  E09000012               Hackney   

   Average_Public_Transport_Accessibility_score,_2014  \
26                                                3.1   
17                                                  3   
36                                                  0   
7                                                 3.2   
13                                                4.3   
29                                                  5   
11                                                4.9   

   Crime_rates_per_thousand_population_2014/15  \
26                                        56.3   
17                                        79.2   
36                                        65.7   
7                                           77   
13                                        90.2   
29                                        99.9   
11                                        99.6   

    Life_satisfaction_score_2011-14_(out_of_10)  \
26                                          7.5   
17                                          7.3   
36                                          7.5   
7                                           7.1   
13                                          7.2   
29                                          7.2   
11                                          7.0   

    Worthwhileness_score_2011-14_(out_of_10)  \
26                                       7.8   
17                                       7.6   
36                                       7.7   
7                                        7.6   
13                                       7.5   
29                                       7.4   
11                                       7.3   

    Happiness_score_2011-14_(out_of_10)  Labels   Latitude  Longitude  
26                                  7.3       0  51.480210  -0.237180  
17                                  7.4       0  51.467220  -0.366530  
36                                  7.3       0  51.546459  -0.161312  
7                                   7.2       0  51.593209  -0.083390  
13                                  7.2       1  51.589264  -0.106405  
29                                  7.2       1  51.520220  -0.054310  
11                                  7.0       1  51.545050  -0.055320


## **Obsevations:All the areas in the city of London ,England have been clustured into three regions as shown above in the map .**

## They have been clustured based on the ***happiness index,worthwhileness index ,lifesatisfaction*** indices that give the over all*** living index*** of these areas . 

The living indices of the clusturs as known from lon_data_new_mean gives us 

# 0	 =7.380769 ->
***BEST PLACE TO START LIVING IN THE CITY OF LONDON ***
#       1	=7.315152--->
**SOME WHAT LESSER SATISFACTION THAN THE CLUSTER 0**
#    2	=7.300000---->
**BETTER CHOOSE FROM FIRST TWO **.

# ***These clustures helps us to choose our living neighbourhood for higher living index**

This is very much needed infromation for almost every businesses that rely on human health and human wellbeing .

for eg:if people are healthy and happy ,there is more likely less crime rate so less no police personell required ,more schools ,more houses in demand in such areas ,more day care centres to name a few .

this is useful for mutiple businesses .



.

```



In [137]:
#using foursquare api .

# define Foursquare Credentials and Version
CLIENT_ID = 'ZSJ2X2HUCASUE3A5U11PJYZJMBLUQMRMVLYMDAWQQYLA2SD1' # your Foursquare ID
CLIENT_SECRET = '44ERGRE0VVSS4VVQCV231CPWIGIQQYI1AGV2EAXLXLLFBVBF' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ZSJ2X2HUCASUE3A5U11PJYZJMBLUQMRMVLYMDAWQQYLA2SD1
CLIENT_SECRET:44ERGRE0VVSS4VVQCV231CPWIGIQQYI1AGV2EAXLXLLFBVBF


In [0]:
#Now, let's get the top 100 venues that are within a radius of 2000 meters.#
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(area_name['Latitude'], area_name['Longitude'], area_name['Area_name']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [139]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Area_name', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()

(3559, 7)


Area_name  Latitude  Longitude              VenueName  VenueLatitude  \
0  City of London   51.5205   -0.09743        Barbican Centre      51.519855   
1  City of London   51.5205   -0.09743   Barbican Art Gallery      51.519800   
2  City of London   51.5205   -0.09743  Barbican Concert Hall      51.519632   
3  City of London   51.5205   -0.09743       Barbican Theatre      51.519722   
4  City of London   51.5205   -0.09743   Barbican Cinemas 2&3      51.521073   

   VenueLongitude          VenueCategory  
0       -0.093942  Performing Arts Venue  
1       -0.093969            Art Gallery  
2       -0.093661           Concert Hall  
3       -0.093606                Theater  
4       -0.092979    Indie Movie Theater

In [141]:
venues_df.groupby(["Area_name"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Area_name                                                               
Barking and Dagenham          41         41         41             41   
Barnet                        42         42         42             42   
Bexley                        81         81         81             81   
Brent                         90         90         90             90   
Bromley                       58         58         58             58   
Camden                       100        100        100            100   
City of London               100        100        100            100   
Croydon                      100        100        100            100   
Ealing                       100        100        100            100   
Enfield                      100        100        100            100   
England                      100        100        100            100   
Greenwich                    100        100        100            100   
Hackney                      100        100        100            100   
Hammersmith and Fulham       100        100        100            100   
Haringey                     100        100        100            100   
Harrow                       100        100        100            100   
Havering                     100        100        100            100   
Hillingdon                   100        100        100            100   
Hounslow                      87         87         87             87   
Inner London                 100        100        100            100   
Islington                    100        100        100            100   
Kensington and Chelsea       100        100        100            100   
Kingston upon Thames         100        100        100            100   
Lambeth                      100        100        100            100   
Lewisham                     100        100        100            100   
London                       100        100        100            100   
Merton                       100        100        100            100   
Newham                        93         93         93             93   
Outer London                 100        100        100            100   
Redbridge                    100        100        100            100   
Richmond upon Thames         100        100        100            100   
Southwark                    100        100        100            100   
Sutton                       100        100        100            100   
Tower Hamlets                100        100        100            100   
United Kingdom               100        100        100            100   
Waltham Forest                67         67         67             67   
Wandsworth                   100        100        100            100   
Westminster                  100        100        100            100   

                        VenueLongitude  VenueCategory  
Area_name                                              
Barking and Dagenham                41             41  
Barnet                              42             42  
Bexley                              81             81  
Brent                               90             90  
Bromley                             58             58  
Camden                             100            100  
City of London                     100            100  
Croydon                            100            100  
Ealing                             100            100  
Enfield                            100            100  
England                            100            100  
Greenwich                          100            100  
Hackney                            100            100  
Hammersmith and Fulham             100            100  
Haringey                           100            100  
Harrow                             100            100  
Havering                           100            100  
Hillingdon                         100 

In [142]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:50]

array(['Performing Arts Venue', 'Art Gallery', 'Concert Hall', 'Theater',
       'Indie Movie Theater', 'Gym / Fitness Center', 'Bar',
       'Street Food Gathering', 'Italian Restaurant', 'Hotel',
       'Vietnamese Restaurant', 'Salad Place', 'Park', 'Sushi Restaurant',
       'Falafel Restaurant', 'Beer Bar', 'Pub', 'Breakfast Spot',
       'Coffee Shop', 'Roof Deck', 'Whisky Bar', 'Grocery Store',
       'Cocktail Bar', 'Flea Market', 'Event Space', 'Steakhouse',
       'Boxing Gym', 'Distillery', 'Butcher', 'Souvlaki Shop', 'Market',
       'Sports Bar', 'Middle Eastern Restaurant', 'Bookstore', 'Spa',
       'Bike Shop', 'Japanese Restaurant', 'Pizza Place',
       'Paper / Office Supplies Store', 'Circus School',
       'Argentinian Restaurant', 'Canal', 'Speakeasy', 'Boutique', 'Café',
       'Mini Golf', 'History Museum', 'Hotel Bar', 'Art Museum',
       'Scenic Lookout'], dtype=object)

In [149]:
# check if the results contain "Shopping Mall"
"Bar" in venues_df['VenueCategory'].unique()

True

In [147]:
# one hot encoding
kl_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
kl_onehot['Area_name'] = venues_df['Area_name'] 

# move neighborhood column to the first column
fixed_columns = [kl_onehot.columns[-1]] + list(kl_onehot.columns[:-1])
kl_onehot = kl_onehot[fixed_columns]

print(kl_onehot.shape)
kl_onehot.head()

(3559, 286)


Area_name  Afghan Restaurant  African Restaurant  Airport Terminal  \
0  City of London                  0                   0                 0   
1  City of London                  0                   0                 0   
2  City of London                  0                   0                 0   
3  City of London                  0                   0                 0   
4  City of London                  0                   0                 0   

   American Restaurant  Antique Shop  Arepa Restaurant  \
0                    0             0                 0   
1                    0             0                 0   
2                    0             0                 0   
3                    0             0                 0   
4                    0             0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            1           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  BBQ Joint  Bagel Shop  \
0                 0                   0            0          0           0   
1                 0                   0            0          0           0   
2                 0                   0            0          0           0   
3                 0                   0            0          0           0   
4                 0                   0            0          0           0   

   Bakery  Bar  Beach  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0       0    0      0                0         0            0           0   
1       0    0      0                0         0            0           0   
2       0    0      0                0         0            0           0   
3       0    0      0                0         0            0           0   
4       0    0      0                0         0            0           0   

   Bike Shop  Bistro  Bookstore  Boutique  Bowling Alley  Boxing Gym  \
0          0       0          0         0              0           0   
1          0       0          0         0              0           0   
2          0       0          0         0              0           0   
3          0       0          0         0              0           0   
4          0       0          0         0              0           0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridge  Bubble Tea Shop  \
0                     0               0        0       0                0   
1                     0               0        0       0                0   
2                     0               0        0       0                0   
3                     0               0        0       0                0   
4                     0               0        0       0                0   

   Building  Burger Joint  Burrito Place  Bus Stop  Butcher  Café  \
0         0             0              0         0        0     0   
1         0             0              0         0        0     0   
2         0             0              0         0        0     0   
3         0             0              0         0        0     0   
4         0             0              0         0        0     0   

   Cajun / Creole Restaurant  Campground  Canal  Canal Lock  \
0                          0           0      0           0   
1                          0           0      0           0   
2                          0           0      0           0   
3                          0           0      0           0   
4                          0           0      0           0   

   Caribbean Restaurant  Castle  Caucasian Restaurant  Chaat Place  \
0                     0       0          

In [148]:
#Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


kl_grouped = kl_onehot.groupby(["Area_name"]).mean().reset_index()

print(kl_grouped.shape)
kl_grouped

(38, 286)


Area_name  Afghan Restaurant  African Restaurant  \
0     Barking and Dagenham               0.00                0.00   
1                   Barnet               0.00                0.00   
2                   Bexley               0.00                0.00   
3                    Brent               0.00                0.00   
4                  Bromley               0.00                0.00   
5                   Camden               0.00                0.00   
6           City of London               0.00                0.00   
7                  Croydon               0.00                0.00   
8                   Ealing               0.00                0.00   
9                  Enfield               0.00                0.00   
10                 England               0.00                0.00   
11               Greenwich               0.00                0.00   
12                 Hackney               0.00                0.00   
13  Hammersmith and Fulham               0.00                0.00   
14                Haringey               0.00                0.00   
15                  Harrow               0.00                0.00   
16                Havering               0.00                0.00   
17              Hillingdon               0.00                0.01   
18                Hounslow               0.00                0.00   
19            Inner London               0.00                0.00   
20               Islington               0.01                0.00   
21  Kensington and Chelsea               0.00                0.00   
22    Kingston upon Thames               0.00                0.00   
23                 Lambeth               0.00                0.01   
24                Lewisham               0.00                0.00   
25                  London               0.00                0.00   
26                  Merton               0.00                0.00   
27                  Newham               0.00                0.00   
28            Outer London               0.00                0.00   
29               Redbridge               0.00                0.01   
30    Richmond upon Thames               0.00                0.00   
31               Southwark               0.00                0.00   
32                  Sutton               0.00                0.00   
33           Tower Hamlets               0.00                0.00   
34          United Kingdom               0.00                0.00   
35          Waltham Forest               0.00                0.00   
36              Wandsworth               0.00                0.00   
37             Westminster               0.00                0.00   

    Airport Terminal  American Restaurant  Antique Shop  Arepa Restaurant  \
0           0.000000             0.000000          0.00              0.00   
1           0.000000             0.000000          0.00              0.00   
2           0.000000             0.000000          0.00              0.00   
3           0.000000             0.000000          0.00              0.00   
4           0.000000             0.000000          0.00              0.00   
5           0.000000             0.010000          0.00              0.00   
6           0.000000             0.000000          0.00              0.00   
7           0.000000             0.000000          0.00              0.00   
8           0.000000             0.000000          0.00              0.00   
9           0.000000             0.000000          0.00              0.01   
10          0.000000             0.000000          0.00              0.00   
11          0.000000             0.000000          0.00              0.00   
12          0.000000             0.000000          0.00              0.00   
13          0.000000             0.000000          0.00              0.00   
14          0.000000             0.010000          0.00              0.00   
15          0.000000             0.000000          0.00              0.00   
16          0.00

In [150]:
len(kl_grouped[kl_grouped["Bar"] > 0])

24

In [168]:
lon_bar = kl_grouped[["Area_name","Bar"]]
lon_bar.head(37)

Area_name       Bar
0     Barking and Dagenham  0.000000
1                   Barnet  0.000000
2                   Bexley  0.024691
3                    Brent  0.011111
4                  Bromley  0.000000
5                   Camden  0.000000
6           City of London  0.020000
7                  Croydon  0.030000
8                   Ealing  0.010000
9                  Enfield  0.010000
10                 England  0.000000
11               Greenwich  0.040000
12                 Hackney  0.010000
13  Hammersmith and Fulham  0.030000
14                Haringey  0.020000
15                  Harrow  0.010000
16                Havering  0.000000
17              Hillingdon  0.040000
18                Hounslow  0.000000
19            Inner London  0.000000
20               Islington  0.010000
21  Kensington and Chelsea  0.010000
22    Kingston upon Thames  0.010000
23                 Lambeth  0.010000
24                Lewisham  0.010000
25                  London  0.000000
26                  Merton  0.040000
27                  Newham  0.021505
28            Outer London  0.000000
29               Redbridge  0.040000
30    Richmond upon Thames  0.020000
31               Southwark  0.000000
32                  Sutton  0.000000
33           Tower Hamlets  0.020000
34          United Kingdom  0.000000
35          Waltham Forest  0.014925
36              Wandsworth  0.030000

In [154]:
# set number of clusters
kclusters = 3

kl_clustering = lon_bar.drop(["Area_name"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(kl_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 0, 0, 1, 1, 0, 2, 0, 0], dtype=int32)

In [0]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
kl_merged = lon_bar.copy()

# add clustering labels
kl_merged["Cluster Labels"] = kmeans.labels_

In [156]:
kl_merged.rename(columns={"Area_name": "Neighborhood"}, inplace=True)
kl_merged.head()

Neighborhood       Bar  Cluster Labels
0  Barking and Dagenham  0.000000               1
1                Barnet  0.000000               1
2                Bexley  0.024691               0
3                 Brent  0.011111               0
4               Bromley  0.000000               1

In [159]:
kl_merged = kl_merged.join(area_name.set_index('Area_name'), on="Neighborhood")

print(kl_merged.shape)
kl_merged.head() # check the last columns!

(38, 5)


Neighborhood       Bar  Cluster Labels   Latitude  Longitude
0  Barking and Dagenham  0.000000               1  51.543932   0.133157
1                Barnet  0.000000               1  51.627294  -0.253759
2                Bexley  0.024691               0  51.622832  -0.080656
3                 Brent  0.011111               0  51.609776  -0.194679
4               Bromley  0.000000               1  51.431820  -0.016566

In [160]:
kl_merged.sort_values(["Cluster Labels"], inplace=True)
kl_merged

Neighborhood       Bar  Cluster Labels   Latitude  Longitude
30    Richmond upon Thames  0.020000               0  51.480210  -0.237180
20               Islington  0.010000               0  51.532790  -0.106140
23                 Lambeth  0.010000               0  51.490840  -0.111080
24                Lewisham  0.010000               0  51.459160  -0.012130
27                  Newham  0.021505               0  51.517368   0.022979
15                  Harrow  0.010000               0  51.513180  -0.106980
14                Haringey  0.020000               0  51.589264  -0.106405
12                 Hackney  0.010000               0  51.545050  -0.055320
21  Kensington and Chelsea  0.010000               0  51.510380  -0.331470
22    Kingston upon Thames  0.010000               0  51.411240  -0.291050
8                   Ealing  0.010000               0  51.514060  -0.300730
6           City of London  0.020000               0  51.520500  -0.097430
33           Tower Hamlets  0.020000               0  51.520220  -0.054310
3                    Brent  0.011111               0  51.609776  -0.194679
2                   Bexley  0.024691               0  51.622832  -0.080656
35          Waltham Forest  0.014925               0  51.581761  -0.276969
9                  Enfield  0.010000               0  51.540021  -0.077501
31               Southwark  0.000000               1  51.505410  -0.089210
32                  Sutton  0.000000               1  51.490987  -0.167417
28            Outer London  0.000000               1  51.524190  -0.157046
34          United Kingdom  0.000000               1  51.506420  -0.127210
25                  London  0.000000               1  51.506420  -0.127210
0     Barking and Dagenham  0.000000               1  51.543932   0.133157
18                Hounslow  0.000000               1  51.467220  -0.366530
16                Havering  0.000000               1  51.544605  -0.144105
10                 England  0.000000               1  51.546459  -0.161312
5                   Camden  0.000000               1  51.532360  -0.127960
4                  Bromley  0.000000               1  51.431820  -0.016566
1                   Barnet  0.000000               1  51.627294  -0.253759
19            Inner London  0.000000               1  51.526179  -0.153859
37             Westminster  0.000000               1  51.500080  -0.128020
36              Wandsworth  0.030000               2  51.456820  -0.194520
17              Hillingdon  0.040000               2  51.484225  -0.096480
26                  Merton  0.040000               2  51.415640  -0.191420
13  Hammersmith and Fulham  0.030000               2  51.482690  -0.212910
29               Redbridge  0.040000               2  51.475778  -0.080698
11               Greenwich  0.040000               2  51.484540   0.002750
7                  Croydon  0.030000               2  51.593209  -0.083390

In [161]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(kl_merged['Latitude'], kl_merged['Longitude'], kl_merged['Neighborhood'], kl_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Observations :from the clusters above it is best evident to start a bar in the cluster 2 as listed as less no of bars are availablein this  cluster .


In [163]:
kl_merged.loc[kl_merged['Cluster Labels'] == 2]

Neighborhood   Bar  Cluster Labels   Latitude  Longitude
36              Wandsworth  0.03               2  51.456820  -0.194520
17              Hillingdon  0.04               2  51.484225  -0.096480
26                  Merton  0.04               2  51.415640  -0.191420
13  Hammersmith and Fulham  0.03               2  51.482690  -0.212910
29               Redbridge  0.04               2  51.475778  -0.080698
11               Greenwich  0.04               2  51.484540   0.002750
7                  Croydon  0.03               2  51.593209  -0.083390